mitiq/shadows/quantum_processing.py

```diff
@@ -0,0 +1,181 @@
+from typing import Tuple, Callable, Dict, Any, List, Union
+
+import cirq
+import numpy as np
+from numpy.typing import NDArray
+from qiskit_aer import Aer
+from tqdm.auto import tqdm
+
+from mitiq import MeasurementResult
+from mitiq.interface.mitiq_cirq.cirq_utils import (
+    sample_bitstrings as cirq_sample_bitstrings,
+)
+from mitiq.interface.mitiq_qiskit.conversions import to_qiskit
+from mitiq.interface.mitiq_qiskit.qiskit_utils import (
+    sample_bitstrings as qiskit_sample_bitstrings,
+)
+
+
+# Functions to implement random Pauli measurements.
+
+
+def generate_random_pauli_strings(
+    num_qubits: int, num_strings: int
+) -> List[str]:
+    """Generate a list of random Pauli strings.
+
+    Args:
+        num_qubits: The number of qubits in the Pauli strings.
+        num_strings: The number of Pauli strings to generate.
+
+    Returns:
+        A list of random Pauli strings.
+    """
+
+    # Sample random Pauli operators uniformly from X, Y, Z
+    unitary_ensemble = ["X", "Y", "Z"]
+    paulis = np.random.choice(unitary_ensemble, (num_strings, num_qubits))
+    return ["".join(pauli) for pauli in paulis]
+
+
+def get_rotated_circuits(
+    circuit: cirq.Circuit,
+    pauli_strings: List[str],
+    add_measurements: bool = True,
+) -> List[cirq.Circuit]:
+    """Returns a list of circuits that are identical to the input circuit,
+    except that each one has single-qubit Clifford gates followed by
+    measurement gates that are designed to measure the input
+    Pauli strings in the Z basis.
+
+    Args:
+        circuit: The circuit to measure.
+        pauli_strings: The Pauli strings to measure in each output circuit.
+        add_measurements: Whether to add measurement gates to the circuit.
+
+    Returns:
+         The list of circuits with rotation and measurement gates appended.
+    """
+    qubits = list(circuit.all_qubits())
+    num_qubits = len(qubits)
+    rotated_circuits = []
+    for pauli_string in pauli_strings:
+        assert len(pauli_string) == num_qubits, (
+            f"Pauli string must be same length as number of qubits, "
+            f"got {len(pauli_string)} and {num_qubits}"
+        )
+        rotated_circuit = circuit.copy()
+        for i, pauli in enumerate(pauli_string):
+            qubit = qubits[i]
+            # Pauli X measurement is equivalent to H plus a Z measurement
+            if pauli == "X":
+                rotated_circuit.append(cirq.H(qubit))
+            # Pauli X measurement is equivalent to S^-1*H plus a Z measurement
+            elif pauli == "Y":
+                rotated_circuit.append(cirq.S(qubit) ** -1)
+                rotated_circuit.append(cirq.H(qubit))
```

Comment:
Could we use something similar to the `PAULI_MAP` dictionary you defined in [#1908](https://github.com/unitaryfund/mitiq/pull/1908/files#diff-0fdf554fbe706eabb33bdc94f895a9b0d3be06b830c272a06dd9f5eb4d6d54b6R11) to simplify this a bit?

```suggestion
        for qubit, pauli in zip(qubits, pauli_string):
            rotated_circuit.append(PAULI_MAP[pauli].on(qubit)
```

Reply:
The `pauli=="Y"` case is not suitable for this style, since it needs two `append` operations in a sequence.

Reply:
You could define a custom gate (the product of $S^\dagger$ and $H$ in this case) using cirq's [`MatrixGate`](https://quantumai.google/reference/python/cirq/MatrixGate). Just thinking of ways to remove redundancy since we have very similar code in another PR.

Reply:
Matrix gate will get this more complicated. A good thing about this format is that when the Pauli is `Z`, no gate is needed to be added.

Reply:
Okay, in that case, the `zip` function is still useful in this code. You can use that to clean up the loop / remove the need for `enumerate`.